### ISSUE TITLE
Loading indicator

### ISSUE THREAD

**Issue by rxng** [r0x0r/pywebview#90]
Hi 

Is there a way to add a loading indicator or possibly even a progress bar to this?

R

**Comment by r0x0r** [comment]
To clarify the question, do you mean a loading indicator to be displayed before WebView is loaded?

**Comment by rxng** [comment]
Yes, that is right.

**Comment by r0x0r** [comment]
Definitely doable, but it is a fairly big effort, as the loading indicator
must be implemented separately for each platform. Creating an animation in
Win32 or Cocoa doesn't sound very appealing.

**Comment by r0x0r** [comment]
I have given it a more thorough thought and imo the way to do it would be introducing two new parameters to the create_window function: background_color and placeholder_image. The first is the background color of the container and the latter would be a path to an image file displayed during loading (with gif support). 
Help is welcomed.

**Comment by shivaprsdv** [comment]
I believe what you just described👆 is an HTML5/CSS3 thing. This is something
the web dev should implement themselves, should they need such eye-candy.
There are any number of websites that does this: load a lightweight page with a
progress indicator, load the main page in the background using AJAX, and switch
once load is finished. Use HTML5 `<progress>` element, jQuery, etc.

What @rxng asked is, I believe, a progress bar like the one shown by browsers,
that would animate on each load. But that would spoil the native-application UX.
So IMO this should not be a part of the API. If web devs intend on loading heavy
pages/internet resources, and want to avoid users staring at blank pages/stuck
pages, they should handle it themselves.

_My two cents._ 🙂

**Comment by r0x0r** [comment]
The loading indicator in question is intended to be displayed before webview is loaded. Currently it a  white background is displayed for a second or two before any HTML content is displayed. Giving an ability to display an image as a loading indicator would make possible to display the same image before the webview is loaded and again the same image in the application itself.

**Comment by shivaprsdv** [comment]
What about, display the window only after the content is fully loaded?

I think this can be implemented without much effort in the source.

**Comment by r0x0r** [comment]
This could do the job, but in this case a loading splash screen would be needed. Progress indication is always better than just leaving the user hanging.

**Comment by shivaprsdv** [comment]
Correct. At least on Mac this shouldn't be a problem, there is the default
'dock-icon-jumping' thing.

**Comment by r0x0r** [comment]
I have been thinking the typical case would be the background colour of the application and a loading animation (a spinner or such). Anyhow both background colour and image would be left up to the user (no defaults). 

Good point about Mac. Hiding the application window until fully loaded could be the default behaviour on Mac.

**Comment by shivaprsdv** [comment]
I take it back. Mac doc-icon jumps only till the application finishes launching,
even if the window is hidden. The webview may take a bit more time to load,
so if we don't show the window, the user is indeed left hanging. \*sigh*

But I have another neat little idea: why not use the *mouse pointer* aka *cursor*
to indicate progress? Those cursors with a spinning wheel/hourglass next to
the arrow? Of course not the ones that indicate that the app is hanging/has
crashed, but these:

![cursor](https://cloud.githubusercontent.com/assets/18247990/25854496/cd13c8a0-34ed-11e7-93e6-8b1f282cea94.png)
These are the cursors shown by browsers for the CSS `cursor: progress;` rule
on each platform.

We just need to give the users visual feedback for a few seconds; they seem to
be perfect for the job. Just hold back the window, show one of these, and when
the webview finishes loading, revert the cursor to normal and show the window.

I've been researching how to do it on my Mac last couple of days. It was trickier
than I thought, but I'm almost there. It seems to be easier to do on Windows and
Linux. I should say this makes our lives a lot easier, for the effort needed for other
solutions that we may come up with.

**Comment by r0x0r** [comment]
The cursor is too subtle on its own, although I have nothing against having it as a complimentary feature. I would like to go with the original approach: customizable background color and an arbitrary centered image or gif

**Comment by shivaprsdv** [comment]
Okay, let us try. BTW I assume you mean to have this based on the`master` branch
for now, is that correct? For I've been hacking on it on top of the multiple-windows
branch, so the concerns were somewhat different.

I'll post the updates if I'm able to do it on Mac. Cheers.

P.S. Here's a small and crude screen cast of my current 'progress' (obviously this is
not what you want, but have a look):

![loading](https://cloud.githubusercontent.com/assets/18247990/25966153/cf345472-36a7-11e7-9e6d-97473087fab4.gif)
_(IDK why, but the spinning of the wheel was not captured)_

**Comment by shivaprsdv** [comment]
Is this👇 what we want?

![loading_latest](https://cloud.githubusercontent.com/assets/18247990/26052310/a95f6a42-3982-11e7-9910-06fee09e4738.gif)

If OK, I shall submit a PR. Can you create a `loading-indicator` branch?
Or should I PR to `master`?

**Comment by r0x0r** [comment]
Exactly. As long as loading animation (file path as a string) and background color (hex color as a string prepended with a hash, eg, "#FFFFFF") can be passed as parameters.
I have created a loading-indicator branch with a create_window parameter signature

**Comment by shivaprsdv** [comment]
Awesome. File-path is relative to the main script. Went a little bit further
with the background color. Users can specify any valid HTML/CSS color
(not just hex, but color-names, `rgb()`, etc).

See PR. 🙂

**Comment by r0x0r** [comment]
I've added a background_color parameter support for Windows. Let's put image support on hold for now.

**Comment by shivaprsdv** [comment]
Agreed. In fact I did this natively in Cocoa at first. I ran into roadblocks while
trying to add GIF image support, and that was why I abandoned the native
approach and resorted to the idea of a placeholder webview.

I have modified and reopened the last PR.

**Comment by shivaprsdv** [comment]
Okay, I've searched enough, still unable to clear the GIF support roadblocks.
Here's what I have:

- Image support can be easily added in Cocoa with an `NSImageView`.
- `NSImageView` supports popular image formats, including GIF.
- The only problem is, it supports only a maximum frame rate of 10 FPS. Images
running at faster frame rates are slowed down.

Should I go ahead and implement it? I hate the frame rate limitation, but something
is better than nothing...

**Comment by r0x0r** [comment]
10FPS might be enough for a simple loading animation, but let's implement a background color first and decide later if a loading animation is needed or not

**Comment by r0x0r** [comment]
I have replaced a bg_color example with a loading_animation example. Indeed at least with this example a "white screen" delay on OSX is very short, so a placeholder animation would not be needed. I will test on how it performs on Windows later.

**Comment by r0x0r** [comment]
Tested with Windows as well. The delay is considerably longer than on OSX, but the background color plus the following CSS animation make the loading process smooth enough.

@rxng would you test loading_animation.py in the loading-indicator branch and tell what you think?

**Comment by shivaprsdv** [comment]
Okay, saw the example. So the idea is to let the user do their own loading
animation in HTML/CSS (may be using the example as a boiler plate) within
their app, right? Well, IIRC this is exactly what I [said before](https://github.com/r0x0r/pywebview/issues/90#issuecomment-299439051). :stuck_out_tongue_winking_eye: 

Bingo, let's do the `background_color` on all platforms and call it a day. :+1:

**Comment by shivaprsdv** [comment]
Updated macOS support. `loading_animation.py` working as expected.

**Comment by shivaprsdv** [comment]
Added support on Linux (GTK).

**Comment by shivaprsdv** [comment]
Added support on Linux (QT).

This leaves only Win32. Any plans to add support on Win32?

BTW when it's ready, can you do a proper merge to `master`?. Last time you
did a sqaush-and-merge (the `disable-backspace` branch) to `master`, I didn't
receive any credit for my contribution. :wink:

**Comment by r0x0r** [comment]
Thanks once again.
I have no plans adding any new features to Win32. I will refine the WinForms implementation (short hex color syntax support should be added for one) and then it can be merged into master.

I mention contributors in the release log for each new release. If you prefer I could merge without squashing, this will preserve commit history in the master branch.

**Comment by shivaprsdv** [comment]
Yes, I think that's the way to go: rebase PR's, but merge feature branches. But
trivial commits can be squashed. Of course it's your project, so it's upto you.:slightly_smiling_face: 
Anyway it's my pleasure to help. Cheers!

Back to business, we should also remove the `loading_image` parameter before
merging.

**Comment by r0x0r** [comment]
The loading-indicator branch is now merged. I have moved a hex color validation out of the platform files into the init and introduced a ValueError exception if the color format is incorrect. 

Closing this issue now. Thanks again for your input.

**Comment by shivaprsdv** [comment]
Good move. Cheers! 🥂
